packages/app-builder/src/hooks/useCursorPaginatedFetcher.ts

```diff
@@ -0,0 +1,72 @@
+import { type SerializeFrom } from '@remix-run/node';
+import { useFetcher } from '@remix-run/react';
+import qs from 'qs';
+import { useCallback, useEffect, useState } from 'react';
+
+import { useCursorPagination } from './useCursorPagination';
+
+type BaseUseCursorPaginatedFetcherOptions<D> = {
+  initialData: D;
+  getQueryParams?: (cursor: string | null) => Record<string, unknown>;
+  validateData?: (data: D) => boolean;
+};
+
+type UseCursorPaginatedFetcherOptions<T, D> =
+  BaseUseCursorPaginatedFetcherOptions<D> &
+    (D extends T
+      ? Record<string, never>
+      : {
+          transform: (value: SerializeFrom<T>) => D;
+        });
+
+export const useCursorPaginatedFetcher = <T, D = T>({
+  initialData,
+  getQueryParams,
+  validateData,
+  ...opts
+}: UseCursorPaginatedFetcherOptions<T, D>) => {
+  const [previousFetcherData, setPreviousFetcherData] =
+    useState<SerializeFrom<T> | null>(null);
+
+  const [data, setData] = useState(initialData);
+  const updateData = useCallback((data: D) => {
+    setData(data);
+  }, []);
```

Comment:
I think you don't need this, use `setData` directly

Reply:
Oh yeah, I don't need it at all, I had other things in there and I didn't notice there was only the setState inside.